typing/typecore.ml

```diff
@@ -7356,6 +7368,35 @@ and type_expect_
             exp_type = exp2.exp_type;
             exp_attributes = sexp.pexp_attributes;
             exp_env = env }
+  | Pexp_quotation exp ->
+      let expected_mode' = expected_mode in
+      let jkind = Jkind.Builtin.value ~why:Quotation_result in
+      let new_env = Env.add_quotation_lock env in
+      let ty = newgenvar jkind in
+      let quoted_ty = newgenty (Tquote ty) in
+      let to_unify = Predef.type_code quoted_ty in
+      with_explanation (fun () ->
+        unify_exp_types loc env to_unify (generic_instance ty_expected));
+      let arg = type_expect new_env expected_mode' exp (mk_expected ty) in
+      re {
+        exp_desc = Texp_quotation arg;
+        exp_loc = loc; exp_extra = [];
+        exp_type = instance ty_expected;
+        exp_attributes = sexp.pexp_attributes;
+        exp_env = new_env }
+  | Pexp_splice exp ->
+      if Env.without_open_quotations env then
+        raise (Error (exp.pexp_loc, env, Toplevel_splice));
+      let argument_mode = mode_legacy in
+      let new_env = Env.add_splice_lock env in
```

Comment:
If we do what I suggested for `add_splice_lock`, here we can simply call `enter_splice`. Then we don't need the function `without_open_quotations`.